org.eclipse.xtext.common.types.ui/src/org/eclipse/xtext/common/types/ui/notification/JavaBuilderState.java

```diff
@@ -0,0 +1,259 @@
+				getSimplePrimaryTypeName(compilationUnit), compilationUnit.getJavaProject());
+	}
+
+	private IPackageFragmentRoot getPackageFragmentRoot(IJavaElement javaElement) {
+		IJavaElement parent = javaElement.getParent();
+		return parent instanceof IPackageFragmentRoot ? ((IPackageFragmentRoot) parent) : getPackageFragmentRoot(parent);
+	}
+
+	private TypeNames getQualifiedTypeNames(String typeLocator, String packageName, String simpleName, IJavaProject project) {
+		TypeNames qualifiedTypeNames = new TypeNames(project);
+		String primaryTypeFqn = getQualifedTypeName(packageName, simpleName);
+
+		char[][] typeNames = null;
+		if (state != null) {
+			typeNames = state.getDefinedTypeNamesFor(typeLocator);
+		}
+
+		if (typeNames == null) {
+			TypeNames newTypeNames = new TypeNames(project);
+			newTypeNames.addTypeName(primaryTypeFqn, primaryTypeFqn);
+			return newTypeNames;
+		}
+
+		for (char[] typeName : typeNames) {
+			String qualifiedTypeName = getQualifedTypeName(packageName, new String(typeName));
+			qualifiedTypeNames.addTypeName(qualifiedTypeName, primaryTypeFqn);
+		}
+
+		return qualifiedTypeNames;
+	}
+
+	private String getQualifedTypeName(String packageName, String simpleTypeName) {
+		return packageName == null ? simpleTypeName : packageName + "." + simpleTypeName;
+	}
+
+	private String getTypeLocator(ICompilationUnit compilationUnit) {
+		IResource resource = compilationUnit.getResource();
+		if (resource != null && resource.getProjectRelativePath() != null) {
+			return resource.getProjectRelativePath().toString();
+		}
+		return null;
+	}
+
+	private String getPackageName(ICompilationUnit compilationUnit) {
+		IJavaElement parent = compilationUnit.getParent();
+		if (parent instanceof IPackageFragment && !((IPackageFragment) parent).isDefaultPackage()) {
+			return ((IPackageFragment) parent).getElementName();
+		}
+		return null;
+	}
+
+	private String getSimplePrimaryTypeName(ICompilationUnit compilationUnit) {
+		String elementName = compilationUnit.getElementName();
+		return elementName.substring(0, elementName.lastIndexOf("."));
+	}
+
+	private SimpleLookupTable getReferences() {
+		if (references != null) {
+			return references;
+		}
+
+		if (state == null) {
+			return null;
+		}
+
+		Object referencesField = readField(state, "references", null);
+		if (referencesField instanceof SimpleLookupTable) {
+			references = (SimpleLookupTable) referencesField;
+		}
+		return references;
+	}
+
+	private Object readField(Object instance, String fieldName, Object defaultValue) {
+		try {
+			Field field = instance.getClass().getDeclaredField(fieldName);
+			field.setAccessible(true);
+			Object value = field.get(instance);
+			if (value != null) {
+				return value;
+			}
+			return defaultValue;
+		} catch (Exception e) {
+			if (LOG.isEnabledFor(Level.ERROR)) {
+				LOG.error(e.getMessage(), e);
+			}
+			return defaultValue;
+		}
+	}
+
+	public TypeNames getQualifiedTypeNames(Object obj) {
```

Comment:
Minor: Could you please move this method up to the declarations of `_getQualifiedTypeNames`?